packages/financial-templates-lib/src/clients/InsuredBridgeL2Client.ts

```diff
@@ -122,14 +127,85 @@ export class InsuredBridgeL2Client {
+      );
+    return eventsData.events[0];
   }
 
-  async getWhitelistTokenEvents(blockSearchConfig: { fromBlock: number; toBlock: number }): Promise<EventData[]> {
-    return await this.bridgeDepositBox.getPastEvents("WhitelistToken", blockSearchConfig);
+  async getWhitelistTokenEvents(blockSearchConfig: BlockSearchConfig): Promise<EventData[]> {
+    const eventsData = await this.getEventsForMultipleProviders(
+      this.fallbackL2Web3s,
+      this.bridgeDepositBoxContract,
+      "WhitelistToken",
+      blockSearchConfig
+    );
+    if (!eventsData.success)
+      throw new Error(
+        `WhitelistToken transaction hash ${eventsData.missingEvent} found in fallback l2 provider not found in first l2 web3 provider`
+      );
+    return eventsData.events[0];
   }
 
+  /**
+   * Fetches specified event data for all input web3 providers. Assumes that the first fetched EventData array contains
+   * the "control" events. Compares the "control" events against each of the other EventData arrays. Returns false
+   * if any of the "control" events are NOT found exactly in ALL of the other EventData arrays.
+   * @param web3s Web3 providers to check for target event.
+   * @param contract Contract to query target event on.
+   * @param eventName Name of target event
+   * @param blockSearchConfig Target event search config
+   * @returns Object containing success of event query, missing event if not found in all providers, and all event data
+   */
+  async getEventsForMultipleProviders(
+    web3s: Web3[],
+    contract: Contract,
+    eventName: string,
+    blockSearchConfig: BlockSearchConfig
+  ): Promise<{ success: boolean; missingEvent: string | null; events: EventData[][] }> {
+    const getEventsPromises = [contract.getPastEvents(eventName, blockSearchConfig)];
+
+    // For each fallback web3 provider, check that the specified events are also found by those providers,
+    // otherwise throw an error. This allows the caller to have additional confidence about the accuracy of fetched
+    // contract state.
+    for (let i = 0; i < web3s.length; i++) {
+      const _contract = new web3s[i].eth.Contract(contract.options.jsonInterface, contract.options.address);
+      getEventsPromises.push(_contract.getPastEvents(eventName, blockSearchConfig));
+    }
+    const events = await Promise.all(getEventsPromises);
+
+    const controlEvents = events[0].map((event) => event.transactionHash);
+
+    // events[0] contains the events returned by the main web3 provider at index 0. We'll compare those events
+    // against those returned by the fallback providers in events[1,...n].
+    for (let i = 1; i < events.length; i++) {
+      const fallbackEvents = events[i];
+      fallbackEvents.forEach((event) => {
+        if (!controlEvents.includes(event.transactionHash)) {
+          return {
+            success: false,
+            missingEvent: event.transactionHash,
```

Comment:
Should we wait to return so we can return a list of _all_ missing events?

Reply:
yup good call

Reply:
I return all missing events now, TBD what we'll do with them but ptal at the interface